File: contestr/pkg/problemcode/problemcode.go
package problemcode

import (
	"fmt"
	"regexp"
	"strconv"
)

var validCode = regexp.MustCompile(`^\d+[A-Z]$`)

// Validate checks regatta problem codes like 1A, 2B.
func Validate(code string) error {
	if !validCode.MatchString(code) {
		return fmt.Errorf("invalid problem_code %q", code)
	}
	return nil
}

// Round extracts tour round number from problem code (e.g. 2A -> 2).
func Round(code string) (int, error) {
	if err := Validate(code); err != nil {
		return 0, err
	}
	i := 0
	for i < len(code) && code[i] >= '0' && code[i] <= '9' {
		i++
	}
	n, err := strconv.Atoi(code[:i])
	if err != nil || n <= 0 {
		return 0, fmt.Errorf("invalid problem_code %q", code)
	}
	return n, nil
}

File: contestr-front/src/shared/utils/problemCode.ts
/** Mirrors contestr/pkg/problemcode (Validate + Round). */

const VALID_CODE = /^\d+[A-Z]$/;

export function isValidProblemCode(code: string): boolean {
    return VALID_CODE.test(code);
}

/** Tour round from code (e.g. "2A" → 2). null if invalid. */
export function roundFromProblemCode(code: string): number | null {
    if (!isValidProblemCode(code)) {
        return null;
    }
    const digits = code.match(/^\d+/)?.[0];
    if (!digits) {
        return null;
    }
    const n = Number.parseInt(digits, 10);
    return n > 0 ? n : null;
}
